templates/etc/init.d/rpm/td-agent

```diff
@@ -209,7 +210,7 @@ do_start() {
   fi
 
   local RETVAL=0
-  daemon --pidfile="${TD_AGENT_PID_FILE}" ${START_STOP_DAEMON_ARGS} "${TD_AGENT_RUBY}" ${TD_AGENT_ARGS} || RETVAL="$?"
+  daemon --pidfile="${TD_AGENT_PID_FILE}" ${START_STOP_DAEMON_ARGS} "${TD_AGENT_RUBY}" ${TD_AGENT_ARGS} \& echo \$! \> "${TD_AGENT_PID_FILE}" || RETVAL="$?"
```

Comment:
Three comments on this:
- `daemon` will "unquote" its arguments anyway (it usually uses `$*`), so you might as well put all of the arguments past `${START_STOP_DAEMON_ARGS}` into one string — it'll make it easier to read and reason about, IMO:
  ```bash
    daemon --pidfile="${TD_AGENT_PID_FILE}" ${START_STOP_DAEMON_ARGS} "${TD_AGENT_RUBY} ${TD_AGENT_ARGS} & echo \$! > ${TD_AGENT_PID_FILE}" || RETVAL="$?"
  ```
- `daemon` uses the return code of the underlying `bash` process as its return code. So, if you background the first command, the return code will always be that of the `echo` command (even if, e.g., `${TD_AGENT_RUBY}` is not found or is not executable). So the only reason it would return a non-zero exit code is if it fails to write the PID file.
  Sadly, the only way to get a return code of a background command is to `wait` for it, but you don't want to `wait` on a successfully backgrounded long-running process. So maybe something like this would work:
  ```bash
  daemon --pidfile="${TD_AGENT_PID_FILE}" ${START_STOP_DAEMON_ARGS} "${TD_AGENT_RUBY} ${TD_AGENT_ARGS} & sleep 0 && kill -0 \$! && echo \$! > ${TD_AGENT_PID_FILE} || wait \$!" || RETVAL="$?"
  ```
  The initial `sleep 0` is needed to give the background command time to fail fast (e.g., if the executable is not available). Tested this one on a few examples, and it seems to do the right thing, but please verify in the context of our script.
- [Optional] Finally, you're unconditionally using `${TD_AGENT_PID_FILE}` here. You could have achieved the same effect by using
  ```bash
  TD_AGENT_ARGS="${TD_AGENT_ARGS} & sleep 0 && kill -0 \$! && echo \$! > ${TD_AGENT_PID_FILE} || wait \$!"
  ```
  in the above `if [ -n "${TD_AGENT_PID_FILE}" ]; then` section, but keeping the current conditional semantics.

Reply:
Just expanding on the second bullet, I understand that this is a fairly esoteric sequence of commands, and that it introduces some technical debt for those unfamiliar with bash scripting. I'm willing to take on that technical debt if necessary.
For posterity, here's how I've tested the commands:
```
$ rm -f /tmp/x; bash -c '/bin/sleep 10 & sleep 0 && kill -0 $! && echo $! >/tmp/x || wait $!'; echo $?; ls -l /tmp/x && (ps -ef | grep $(cat /tmp/x))
0
-rw-r--r-- 1 igorp primarygroup 7 Dec 16 12:39 /tmp/x
igorp    120957      1  0 12:39 pts/14   00:00:00 /bin/sleep 10
igorp    120962 120960  0 12:39 pts/14   00:00:00 grep --color=auto 120957
$ rm -f /tmp/x; bash -c '/bin/sleepp 10 & sleep 0 && kill -0 $! && echo $! >/tmp/x || wait $!'; echo $?; ls -l /tmp/x && (ps -ef | grep $(cat /tmp/x))
bash: /bin/sleepp: No such file or directory
bash: line 0: kill: (121003) - No such process
127
ls: cannot access '/tmp/x': No such file or directory
$ rm -f /tmp/x; bash -c '/usr/sbin/uplink & sleep 0 && kill -0 $! && echo $! >/tmp/x || wait $!'; echo $?; ls -l /tmp/x && (ps -ef | grep $(cat /tmp/x))
bash: /usr/sbin/uplink: Permission denied
bash: line 0: kill: (122035) - No such process
126
ls: cannot access '/tmp/x': No such file or directory
```
Optionally, to quell the "`bash: line 0: kill: (NNNN) - No such process`" warnings, we could add `2>/dev/null` to the `kill` command:
```bash
  daemon --pidfile="${TD_AGENT_PID_FILE}" ${START_STOP_DAEMON_ARGS} "${TD_AGENT_RUBY} ${TD_AGENT_ARGS} & sleep 0 && kill -0 \$! 2>/dev/null && echo \$! > ${TD_AGENT_PID_FILE} || wait \$!" || RETVAL="$?"
```

Reply:
> * [Optional] Finally, you're unconditionally using `${TD_AGENT_PID_FILE}` here. You could have achieved the same effect by using
>   ```shell
>   TD_AGENT_ARGS="${TD_AGENT_ARGS} & sleep 0 && kill -0 \$! && echo \$! > ${TD_AGENT_PID_FILE} || wait \$!"
>   ```
>   
>   
>   in the above `if [ -n "${TD_AGENT_PID_FILE}" ]; then` section, but keeping the current conditional semantics.

This is feasible, but the script already assumes that TD_AGENT_PID_FILE is set in another place, so the change proposed wouldn't affect the behavior.